### DISCUSSION TITLE
Themes

### DISCUSSION THREAD

**Discussion by heff** [muxinc/media-chrome#323] [Proposals]
Prior context: [Theme Architecture PR](https://github.com/muxinc/media-chrome/pull/42). 

There's a handful of threads on themes happening so I've been thinking about it and want to propose a potentially controversial change to the existing themes concept, and surface any clear issues with the direction.

> **Proposal**: Themes are just HTML and CSS, representable as an HTML template tag, or JS template string.

A **Skin** is CSS.
A **Theme** is HTML + CSS
Adding JS to the equation is the realm of the **Player**.

The [original theme examples](https://github.com/muxinc/media-chrome/blob/main/src/js/themes/media-theme-netflix.js#L13) are basically that, with the added wrapper of a custom element to make embedding and registering easier.

With the [mux player theme](https://github.com/muxinc/elements/blob/main/packages/mux-player/src/media-theme-mux/media-theme-mux.ts) we then expanded on this (using javascript and theme element attributes) to support things like:

* [Responsive sizing of the player](https://github.com/muxinc/elements/blob/main/packages/mux-player/src/media-theme-mux/media-theme-mux.ts#L11)
* [Different UIs depending on the stream type](https://github.com/muxinc/elements/blob/main/packages/mux-player/src/media-theme-mux/media-theme-mux.ts#L141)
* (other things I'm missing?)

This proposal ultimately needs to solve for those needs to be complete. I don't have immediate solutions but I have thoughts and I'm interested in others' ideas.

In general I'm drawn towards the simplicity of the concept, but some rough benefits I see with this approach:
* Anyone that can HTML can create and share a theme (i.e. many designers)
* You can easily copy/paste a theme's contents into your own doc and customize it from there
* Themes could be made more _secure_ in any central repository of community-created themes by stripping out any javascript. (Allow anyone to submit a theme. Allow loading any theme directly from the repository hosting.)

Thoughts?

**Comment by mmcc** [comment]
> A Theme is HTML + CSS

Just curious, what about custom elements? Do those count as HTML in this context or would that fall under "adding JS to the equation" (yes, it's literally adding JS, but roll with me).

**Reply by heff** [reply]
The key concept is you'd be creating just a string of HTML, and custom elements are HTML, so yeah, custom elements would work. But they would have to be available in the context that the theme is being used in, they don't just come along with the theme. That would be obvious for media chrome elements. Others probably look like theme dependencies.

**Reply by gkatsev** [reply]
I think that we should lean into the web component angle more and define themes as web components that wrap media-chrome+HTML+CSS. If themes are just HTML/CSS it makes sharing them not as easy, but if they're web components, folks can grab them and use them more easily. This does blur the line a bit between Player and Theme, but Themes probably shouldn't be including much extra functionality that a player could/would. Themes should also then provide enough hooks/slots/parts/etc so that a Player using the theme has access to everything it needs for the added functionality it will add.

**Reply by heff** [reply]
> If themes are just HTML/CSS it makes sharing them not as easy

Can you say more about what you mean by that?

> but Themes probably shouldn't be including much extra functionality that a player could/would

Where could we draw a clear line then?

> Themes should also then provide enough hooks/slots/parts/etc so that a Player using the theme **has access to everything it needs** for the added functionality it will add

This is where I get a little concerned about the future complexity of themes. I'm not sure what the exact needs will be, but if _players_ ultimately want access to more inside the theme, like the media-controller, I feel like it'd be better if players were just cloning theme HTML into their doms, rather than making the Theme element the middle person, and expecting more of themes than just defining the layout.

I do still like using a theme custom element in a simple, non-player situation for easy theming, but that could just be a generic `<media-theme-loader>` element if themes are templates.

**Reply by gkatsev** [reply]
CSS is decently easy to share, but HTML isn't really. Since, with CSS you can `@import` but HTML doesn't have anything like that right now, so, you'd need to copy it over. Personally, while I understand how themes for stuff like 11ty et al are done where you clone the repo and modify it locally, I don't think it's that nice because then it becomes hard to stay up to date with the upstream theme as things are changed there.

@cjpillsbury is writing a longer answer with more details that should cover the functionality aspect of things.

**Comment by luwes** [comment]
I like this concept but more as a Media Chrome build tool that can be used to do rapid theme development but with more constraints, good defaults. 

I had worked on something in these lines a while back, goal was to separate each asset html, css, svg in its own container. https://github.com/luwes/media-theme-jump-start

I'm a little worried if this would be the only way to build themes, loses a lot of flexibility that JS provides.
Do we make choice for the theme developers how to show/hide controls for responsiveness? Removing them from DOM or using CSS.

**Reply by heff** [reply]
> loses a lot of flexibility that JS provides

What if javascript is allowed, but themes are still defined as html (an html doc partial, with a slot for the media)? Include any js and css you want in that html. Does that relieve some of the concern?

The media-theme-jump-start is awesome, and also a little intimidating. At least coming from the world of video.js CSS skins.

> Do we make choice for the theme developers how to show/hide controls for responsiveness?

I do think media-container ultimately has to provide an opinionated solve for responsiveness, but I don't know that it has to be the only way.

**Comment by cjpillsbury** [comment]
> * (other things I'm missing?)

Hey I was hoping to draft a document broadly discussing these topics. The majority of these fall into allowing a theme to be in some sense "configurable"/"dynamic", as opposed to a static, "render once in a very specific way" DOM structure (including attributes). Likewise, if these are missing from a given theme/from a constrained view of "themes," the majority of what will result can be described as "reduced functionality and configurability," and not catastrophic errors. I'll be providing separate replies for the different categories as I draft them so folks can respond to each as a thread.

**Reply by cjpillsbury** [reply]
My S1 is it would be unfortunate if relying on themes resulted in developers/implementors having to make tradeoffs on functionality they get from using Media Chrome directly vs. functionality they get from using a "theme". Unless we come up with alternative solutions (which can/arguably should include "beefing up" `MediaTheme` imo), these would include:

- responsive designs (effectively a stopgap for sufficiently wide [CSS container queries support](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Container_Queries))
- a11y considerations for keyboard navigation and screen readers (e.g. conditionally disabling controls, conditionally disabling hotkeys)
- easy to use configuration attributes (e.g. seek offsets, default captions enabled/disabled, nohotkeys, playback rates, etc.)

**Comment by cjpillsbury** [comment]
### Providing well defined configuration attributes to Media Chrome elements

Things like `forward-seek-offset`, `playback-rates`, `poster` (src for the poster element), `hotkeys`/`nohotkeys` provide attributes down to the `media-controller` and the controls/display elements (and may change over time).

Note that nothing will "catastrophically break" if these simply aren't made available in a "basic theme," but that does mean folks will loose functionality if they decide to adopt a basic theme. While that will always be true to some extent, since these are well-defined on "official" Media Chrome elements, this is something we could architect for.

**Reply by gkatsev** [reply]
Some of these could be provided on the MediaTheme class.

**Reply by cjpillsbury** [reply]
correct/agree, though something would have to be responsible for "detecting" elements in the shadow DOM template and applying the attributes accordingly (by convention) if we want a particular theme to preclude javascript/dynamic updates to their template. So there are two parts:
1. support for well-defined attributes on themes at all and whose responsibility that is
    1. e.g. `observedAttributes` and `attributeChangedCallback` implementations. These could definitely be added to `MediaTheme` in a well-defined/predictable/documented way and also in a way that doesn't necessitate JavaScript in an _**implementation**_ of a particular theme.
3. support for applying said attributes and their changes over time to the shadow DOM elements.
    i. This could also be supported generically in `MediaTheme` using mutation observers to also avoid JavaScript in theme implementations, though it'll be a bit clunkier and may surprise implementors of themes.
    ii. Alternatively, we intentionally made an extendable `render` method on `MediaTheme`. We could update `MediaTheme` to pass down the latest props/attributes (roughly like e.g. React's render), but that would require allowing some level of JavaScript in the HTML.

**Comment by cjpillsbury** [comment]
### Providing attributes that give context/content to rendering the UI

Some of these are simple, such as having a `title` attribute to pass to a `<media-text-display/>`. Things like these could be handled via custom slots in a theme, though that does add greater verbosity/burden on devs using a particular slot. Some of these are state that could come from the slotted media, but is currently un-modeled/unmanaged state in `<media-controller/>`, such as using `has-src` to determine what controls should be disabled, and, arguably, `stream-type`, which at least has [a discussion thread in media-ui-extensions](https://github.com/video-dev/media-ui-extensions/issues/3).

Here, there may be some UI errors or unanticipated/"weird" state if users e.g. interact with the UI before a src is set (there's a related conversation in media chrome here #288). Under most use cases, however, this is similar to "Providing well defined configuration attributes" in that nothing will "catastrophically break" but some features will be unavailable.

**Comment by cjpillsbury** [comment]
### Providing custom elements via slots to externalize components

Currently we rely on two custom slots in our theme: `seek-live` for the seek to live button + indicator, and `mxp-dialog` for our error dialog. For both of these, the primary reason we use slots is because they rely on javascript/information available at the "player" level, such as the state management of errors and the computation of live edge window used by e.g. the component's click handler and the live edge indicator. 

Both were slotted because both the state and the components weren't "well-defined" in Media Chrome in a way that aligned with our timeline. Note that even if these do eventually get integrated into Media Chrome's state and component architecture, we've previously acknowledged the possibility that some things that make sense for Mux Player may never get formally adopted into Media Chrome, and that it's even more plausible that there will at least be a delay between adoption to Media Chrome and need in Mux Player. Here again, nothing will catastrophically break if these are unavailable, but that does mean users and implementors would loose these features when using a "basic" theme that precludes javascript. (Though hypothetically we could refactor the error dialog to not be part of the theme at all, since it's an "overlay".

**Comment by cjpillsbury** [comment]
### `exportparts` and automatically adding `parts`

Currently to facilitate advanced styling easily for devs, we both add an `exportparts` to our usage of our theme (and conceptually any theme) based on well defined components and layers. We also added a simple script that, aside from a small edge case that we could rework, uses slot and component name matchers to [generically compute and set a parts attribute on an element](https://github.com/muxinc/elements/blob/main/packages/mux-player/src/media-theme-mux/media-theme-mux.ts#L108-L132).

As @luwes's code comment suggests, this could be added to the base class of `MediaTheme`. Note that `parts` will be needed even for Mux Player-specific styles/css, including well-defined css variables, for them to be able to "target" any region/element in a given theme.

**Comment by dylanjha** [comment]
> In general I'm drawn towards the simplicity of the concept, but some rough benefits I see with this approach:

++ I like simplicity, especially as a starting point

> Anyone that can HTML can create and share a theme (i.e. many designers)

I think this should be true no matter what. The simplest version of a theme should be exactly what you're proposing here [demuxed 2022 theme](https://github.com/maveio/media-chrome/blob/6c6153dedc1071b9a90988e2bb7b84bd943132cd/src/js/themes/media-theme-demuxed.js).

I think right now, the big blocker to even having a simple theme be pure HTML + CSS is the lack of responsiveness primitives. Even the simplest theme must be able to be responsive.

If we have that, then the simplest themes can be pure HLML + CSS, thus unlocking the ability for any designer who knows HTML + CSS to build a theme without touching javascript

> You can easily copy/paste a theme's contents into your own doc and customize it from there
> Themes could be made more secure in any central repository of community-created themes by stripping out any javascript. (Allow anyone to submit a theme. Allow loading any theme directly from the repository hosting.)

These 2 benefits don't feel super strong to me, loading a custom element or copy/pasting custom element code is just as easy as copy/pasting HTML and CSS

---

I think where this breaks down, and where javascript is required is for any theme to do anything slightly interesting, beyond the basics.

### Example

I am creating a theme that has a "Buy Now" button in the top right of the player, I might define an interface like

```html
<media-theme-shopping buy-now-target="/checkout-page">
  <video slot="media"></video>
</media-theme-shopping>
```

In order to do that, I would necessarily need some JS that listens to the `attributeChangedCallback`, reads the attribute for `buy-now-target`, and sets the target of the button. Maybe for this particular example a theme author would have to use a slot instead?

I would think that theme authors can define attributes like this, how would it be done with JS or without setting up an `attributeChangedCallback` listener?

**Comment by heff** [comment]
I appreciate all the feedback here, and I've been ruminating on all the conversations.

One detail that's come up a few times is _remixing_, and I think it's worth calling out "ease of remixing" as a clear goal, that's on the other end of the spectrum from "advanced/dynamic player". What I mean by _remixing_ is the ability to take an existing theme and modify it completely, whether to be shared as a new theme or just customized within the user's project. **I wouldn't be surprised if this is an incredibly common use case**, compared with using the pre-baked custom element version of a theme, which you can only customize by setting variables. On our future themes site, we don't just want a handful of themes, we want thousands of them, so we should pay close attention to the friction points in creating a theme.

From that angle, I'm wondering if we require advanced/dynamic themes to still expose a [static template](https://github.com/muxinc/media-chrome/blob/main/src/js/themes/media-theme.js#L8) version of itself for at least the purpose of remixing. e.g. for Mux Player this look like exposing the Large On-demand version of the HTML as the static template.

This probably deserves a bigger, clearer writeup, but I'm interested in any initial thoughts on that.

**Reply by luwes** [reply]
I like that idea a lot but not sure how it would work.

So simplified conclusion:

- designer themes: HTML file (mainly HTML + CSS) that users can copy/paste and move around how they see fit
- dev themes: theme JS web component ala  Mux player's `media-theme-mux` 

For the designer themes overcoming that they need to copy / paste the whole theme is the hard part

CSS is easily overridden like videojs but HTML structure is more difficult. slots are the most powerful concept here I think. 
so I could imagine some base / default theme which is responsive by default but the inner controls are all configurable. 
really use HTML as a configuration language almost instead of layout, that can override the defaults but don't have to

```html
<media-default-theme>
	<svg slot="play-icon" path="...." />
	<div slot="extra-small-bottom-controls">
		<media-play-button></media-play-button>
		<my-buy-button></my-buy-button>
	</div>
	<div slot="large-top-right">
		...
	</div>
</media-default-theme>
```

I guess stream type etc would also have to be included somehow.

we should probably do a spike on a "default" responsive and stream type aware Media Chrome theme and try not to force any JS on the end user

**Reply by heff** [reply]
> So simplified conclusion...

Yeah, I think we're on the same page, but after solving responsiveness and stream type at the media controller level, I would expect a lot if not the majority of themes on the themes site to be in the realm of _designer_.

Maybe a better way to frame it is:

### Theme 
A web component (`<media-theme-netflix>`) that expects a slot=media element, can optionally allow other custom slots, and may have attributes for other configuration options. A theme includes a property/function/API to retrieve a _theme template_ version of itself, for deeper customization. This "static" version of the theme may not include everything the theme can do dynamically, e.g. modify the HTML structure based on a variable.

### Theme template
A template string or template element, which includes HTML and CSS, can include `<slot>`s, and can use [template-parts](https://github.com/github/template-parts). A theme template can be easily packaged as a custom element using `<media-theme>` or easily loaded directly using the `<media-theme-template-loader>`, where all custom slots still work and element attributes are translated to template parts.

```html

<!-- Create the theme template -->
<template id="myFancyTheme">
  <media-controller>
    <slot name="media" slot="media"></slot>
    <media-control-bar>
        <media-title>{{ myVideoTitle }}</media-title>
        <slot name="myCustomControl"></slot>        
    </media-control-bar>
  </media-controller>
</template>

<!-- Load the theme template -->
<media-theme-template-loader myVideoTitle="Element attributes translate to template parts!">
    <video slot="media" src=""></video>
    <media-play-button slot="myCustomControl"></media-play-button>
</media-theme-template-loader>
```

> For the designer themes overcoming that they need to copy / paste the whole theme is the hard part...CSS is easily overridden like videojs but HTML structure is more difficult.

I'm not totally following what you mean here. Or at least, I've always felt like HTML was easier than CSS, which I think you're saying the opposite, right?

**Reply by cjpillsbury** [reply]
@heff from our out of band conversation, I think it is a good idea to at least try to have have any "official," "dynamic/templated" theme be able to produce a static representation of what it would/should look like by default/on initial load. I can think of places where this breaks down/puts more burden/"handcuffs" some use cases (e.g. where attribute values are tied to info coming from the slotted media element), but it's a good target to aim for (not sure if it's "blue skies" vs realistic, but that shouldn't matter here).

**Reply by luwes** [reply]
> I'm not totally following what you mean here. Or at least, I've always felt like HTML was easier than CSS, which I think you're saying the opposite, right?

I guess I wanted to say here that CSS is easily overridden, because of CSS specificity I can override rules and chance styles.
I agree HTML is easier but it doesn't have built-in ways to change markup easily by override, full replacement is required here.
maybe a strange idea. but yea with slots you could make layouts or with a templating language.

**Comment by heff** [comment]
With the work @luwes [has been doing on themes](https://github.com/muxinc/media-chrome/pull/362), I'm gonna summarize the current proposal as clearly as I can to invite push back.

Themes are `<template>`s, that use a templating language built on [the template parts proposal](https://github.com/github/template-parts). 

```html
<template id="my-fancy-theme">
        <style>...</style>

        <media-controller>
            <slot name="media" slot="media"></slot>
            <media-text-display>{{ title }}</media-text-display>
            ...
        </media-controller>
</template>
```

Themes may be packaged as a custom element (`<my-fancy-media-theme>`), or loaded by something like a `<media-theme-loader theme="my-fancy-theme">`, both of which will then provide a consistent interface to the theme element, including the ability to set variables, slot elements, and access the mediaController of the theme directly in javascript.

What we're explicitly saying that is different from before is that a theme is not inherently a custom element, it is a `<template>`. That template _can_ be packaged as a custom element for portability, and we will likely make that easy, but that's not a requirement. **A theme starts as, and can always be represented as a `<template>`.**

We do still have more to figure out, but we're feeling confident in this direction.
* What does it look like to _package_ a theme?
  * What if it has dependencies like image files?
  * What if it has dependencies like non-media-chrome custom elements?

We also have things we need to build into media chrome to support common player features without custom code, like seek-to-live.

Also, we acknowledge that themes won't be able to solve everything. We want to ultimately make that a small set of things, but for some advanced features you may have to use media chrome elements directly. With themes as `<template>`s, you can still easily extract the internals of a theme to use as a starting point for your _player_, and build advanced things on to.
